src/test/java/org/cli/CliTests.java

```diff
@@ -0,0 +1,178 @@
+
+/**
+ *
+ * Tests for the CommCare CLI
+ *
+ * Uses a specific, highly paired format to deal with the CLI's I/O
+ *
+ * @author wpride
+ */
+
+public class CliTests {
+
+    private class CliTestRun<E extends CliTestReader> {
+
+        CliTestRun(String applicationPath,
+                          String restoreResource,
+                          Class<E> clazz,
+                          String steps) throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
+            ApplicationHost host = buildApplicationHost(applicationPath, restoreResource, clazz, steps);
+            boolean passed = false;
+            try {
+                host.run();
+            } catch (EarlyExitException e) {
+                passed = true;
+            }
+            assertTrue(passed);
+        }
+
+        private ApplicationHost buildApplicationHost(String applicationPath,
+                                                     String restoreResource,
+                                                     Class<E> clazz,
+                                                     String steps) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+            ClassLoader classLoader = getClass().getClassLoader();
+            File appFile = new File(classLoader.getResource(applicationPath).getFile());
+            String resourcePath = appFile.getAbsolutePath();
+            PrototypeFactory prototypeFactory = new LivePrototypeFactory();
+
+            CommCareConfigEngine engine = CliCommand.configureApp(resourcePath, prototypeFactory);
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            PrintStream outStream = new PrintStream(baos);
+
+            Constructor<E> ctor = clazz.getConstructor(String.class, ByteArrayOutputStream.class);
+            CliTestReader reader = ctor.newInstance(steps, baos);
+
+            ApplicationHost host = new ApplicationHost(engine, prototypeFactory, reader, outStream);
+            File restoreFile = new File(classLoader.getResource(restoreResource).getFile());
+            String restorePath = restoreFile.getAbsolutePath();
+            host.setRestoreToLocalFile(restorePath);
+            return host;
+        }
+    }
+
+    @Test
+    public void testConstraintsForm() throws Exception {
+        // Start a basic form
+        new CliTestRun<>("basic_app/basic.ccz",
+                "case_create_basic.xml",
+                BasicTestReader.class,
+                "1 0 \n");
+    }
+
+    @Test
+    public void testCaseSelection() throws Exception {
+        // Perform case selection
+        new CliTestRun<>("basic_app/basic.ccz",
+                "basic_app/restore.xml",
+                CaseTestReader.class,
+                "2 1 5 \n");
+    }
+
+    static abstract class CliTestReader extends BufferedReader {
+
+        private String[] args;
+        private int index;
```

Comment:
Can you clarify what `index` represents with a more descriptive name and/or comment? It's tough to follow